File: python/all-paths-from-source-to-target.py
class Solution:
    def get_paths(graph: List[List[int]], current_path: List[int], solutions: List[List[int]], target: int):
        tail = current_path[-1]
        if tail == target:
            solutions.append(list(current_path))
        else:
            for new_tail in graph[tail]:
                current_path.append(new_tail)
                Solution.get_paths(graph, current_path, solutions, target)
                current_path.pop()


    def solution_1(graph: List[List[int]]) -> List[List[int]]:
        solutions = []
        current_path = [0]
        target = len(graph) - 1
        Solution.get_paths(graph, current_path, solutions, target)
        return solutions


    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        solutions = []
        current_path = [0]
        target = len(graph) - 1
        Solution.get_paths(graph, current_path, solutions, target)
        return solutions

File: go/all-paths-from-source-to-target.go
func getPaths(graph *[][]int, current_path *[]int, solutions *[][]int, target int) {
    current_path_size := len(*current_path)
    tail := (*current_path)[current_path_size - 1]
    if tail == target {
        solution := make([]int, current_path_size)
        copy(solution, *current_path)
        *solutions = append(*solutions, solution)
    } else {
        for i := 0; i < len((*graph)[tail]); i++ {
            new_tail := (*graph)[tail][i]
            *current_path = append(*current_path, new_tail)
            getPaths(graph, current_path, solutions, target)
            *current_path = (*current_path)[:current_path_size]
        }
    }
}

func allPathsSourceTarget(graph [][]int) [][]int {
    var solutions [][]int
    current_path := make([]int, 1)
    target := len(graph) - 1
    getPaths(&graph, &current_path, &solutions, target)
    return solutions
}
